source/validate_cfg.cpp

```diff
@@ -59,56 +60,72 @@ using bb_ptr = BasicBlock*;
 using cbb_ptr = const BasicBlock*;
 using bb_iter = vector<BasicBlock*>::const_iterator;
 
-/// @brief Sorts the blocks in a CFG given the entry node
-///
-/// Returns a vector of basic block pointers in a Control Flow Graph(CFG) which
-/// are sorted in the order they were accessed in a post order traversal.
-///
-/// @param[in] entry the first block of a CFG
-/// @param[in] depth_hint a hint about the depth of the CFG
-///
-/// @return A vector of pointers in the order they were access in a post order
-/// traversal
-vector<const BasicBlock*> PostOrderSort(const BasicBlock& entry, size_t size) {
-  struct block_info {
-    cbb_ptr block;
-    bb_iter iter;
-  };
+using get_blocks_func = function<const vector<BasicBlock*>*(const BasicBlock*)>;
+
+struct block_info {
+  cbb_ptr block;
+  bb_iter iter;
+};
+
+bool IsBackEdge(vector<block_info> staged, uint32_t id) {
+  for (auto b : staged) {
+    if (b.block->get_id() == id) return true;
+  }
+  return false;
+}
 
+/// @brief Depth first traversal starting from the \p entry BasicBlock
+///
+/// This function performs a depth first traversal from the \p entry
+/// BasicBlock and calls the pre/postorder functions when it needs to process
+/// the node in pre order, post order. It also calls the backedge function
+/// when a back edge is encountered
+void DepthFirstTraversal(const BasicBlock& entry,
+                         get_blocks_func successor_func,
+                         function<void(cbb_ptr)> preorder,
+                         function<void(cbb_ptr)> postorder,
+                         function<void(cbb_ptr, cbb_ptr)> backedge) {
   vector<cbb_ptr> out;
   vector<block_info> staged;
   unordered_set<uint32_t> processed;
 
-  staged.reserve(size);
-  staged.emplace_back(block_info{&entry, begin(entry.get_successors())});
-  processed.insert(entry.get_id());
+  staged.reserve(10);
+  staged.push_back({&entry, begin(*successor_func(&entry))});
+  preorder(&entry);
 
   while (!staged.empty()) {
     block_info& top = staged.back();
-    if (top.iter == end(top.block->get_successors())) {
-      out.push_back(top.block);
+    if (top.iter == end(*successor_func(top.block))) {
+      postorder(top.block);
       staged.pop_back();
     } else {
       BasicBlock* child = *top.iter;
       top.iter++;
-      if (processed.find(child->get_id()) == end(processed)) {
-        staged.emplace_back(block_info{child, begin(child->get_successors())});
+      if (IsBackEdge(staged, child->get_id())) {
```

Comment:
It's ok to replace with a function, as long as the function name (and blurb) describe exactly what the function does.  "IsBackEdge" as a name is poor in this case because that pulls in assumptions about the context in which it's called. It's really "is the given Id in this container?".